offsetbycodepoints function added #1590 (#2453)
Codecademy/docs: content/java/concepts/strings/terms/offsetByCodePoints/offsetByCodePoints.md

---

> Title: 'offsetByCodePoints()'

Reviewer: ```suggestion
Title: '.offsetByCodePoints()'
```

It is useful when working with strings that contain characters with multiple code points, such as emojis or characters from non-Latin scripts. This function helps accurately navigate and manipulate the index based on code points, allowing you to access specific characters or portions of the string.

## Syntax

```java
Strings.offsetByCodePoints(int startIndex,int  offsetIndex)
```

- `startIndex` (int): The starting index in the string from which the offset is applied.

- `OffsetIndex` (int): The offset, in terms of code points, by which the index is adjusted. Positive values move the index forward, and negative values move it backward.

## Example 1

In this example, the code point offset is set to 7, which means the new index will be 7. It moves the index forward by 7 code points, effectively pointing to the character `'W'` in the string.

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
  String str = "Hello, World!";
  int startIndex = 0; // Starting index
  int offsetIndex = 7; // Code point offset
  int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
  System.out.println("New Index: " + newIndex);

>   // Output: New Index: 7

Reviewer: Remove the comment and all the blank lines before the closing brackets.

  }
}
```

The output would be:

```
New Index: 7
```

> ## Example 2

Reviewer: The next two examples are essentially the same as the first one. It might be good to show an example of fetching a character from the middle of a string of completely non-Latin script.

In this example, the string str contains the phrase `"hello, दुनिया!"`(meaning "hello, world!" in English). The starting index is set to 0, and the code point offset is 6.

The `**offsetByCodePoints()**` method calculates the new index by moving it forward by 6 code points. The new index will be 6, pointing to the character `'द'` (da) in the string.

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
  String str = "hello, दुनिया!";
  int startIndex = 0; // Starting index
  int offsetIndex = 6; // Code point offset

  int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
  System.out.println("New Index: " + newIndex);

>   // Output: New Index: 6

Reviewer: Remove the comment and the extra blank lines.

Author: Okay

  }
}
```

The output would be:

```
New Index: 6
```

## Example 3

In this example, the string str contains the text `"Hello, 世界!".` We specify the starting index as 0 and the code point offset as 5. The `**offsetByCodePoints()**` method returns the new index after applying the offset, which is 5 in this case. It means that the character at index 5 is the desired character.

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
   String str = "Hello, 世界!";
   int startIndex = 0; // Starting index
   int offsetIndex = 5; // Code point offset

   int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
   System.out.println("New Index: " + newIndex);

>    // Output: New Index: 5

Reviewer: Remove the comment and the extra blank lines.

---
Title: '.offsetByCodePoints()'
Description: 'Returns the new index of a character in a string after applying the specified code point offset.'
Subjects:
  - 'Computer Science'
Tags:
  - 'Strings'
  - 'Methods'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

The **`offsetByCodePoints()`** function is used to calculate the index of a character in a string by specifying a starting index and a code point offset. It returns the new index after applying the offset limit.

It is useful when working with strings that contain characters with multiple code points, such as emojis or characters from non-Latin scripts. This function helps accurately navigate and manipulate the index based on code points, allowing you to access specific characters or portions of the string.

## Syntax

> ```java

Reviewer 2: ```suggestion
```pseudo
```

> Strings.offsetByCodePoints(int startIndex,int  offsetIndex)

Reviewer 2: ```suggestion
Strings.offsetByCodePoints(startIndex, offsetIndex)
```

```

- `startIndex` (int): The starting index in the string from which the offset is applied.

> - `OffsetIndex` (int): The offset, in terms of code points, by which the index is adjusted. Positive values move the index forward, and negative values move it backward.

Reviewer 2: ```suggestion
- `offsetIndex` (int): The offset, in terms of code points, by which the index is adjusted. Positive values move the index forward, and negative values move it backward.
```

## Example 1

> In this example, the code point offset is set to 7, which means the new index will be 7. It moves the index forward by 7 code points, effectively pointing to the character `'W'` in the string.

Reviewer 2: ```suggestion
In this example, the code point offset is set to `7`, which means the new index will be `7`. This results in the new index pointing to the character `'W'` in the string.
```

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
  String str = "Hello, World!";
  int startIndex = 0;
  int offsetIndex = 7;

  int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
  System.out.println("New Index: " + newIndex);
  }
}
```

> The output would be:

Reviewer 2: ```suggestion
The code above results in the following output:
```

```
New Index: 7
```

## Example 2

> In this example, the string str contains the text `"こんにちは、世界！"`. We specify the starting index as 0 and the code point offset as 5. The `**offsetByCodePoints()**` method returns the new index after applying the offset, which is 5 in this case. It means that the character at index 5 is the desired character.

Reviewer 2: ```suggestion
In this example, the string str contains the text `"こんにちは、世界！"`. We specify the starting index as `0` and the code point offset as `5`. The `**offsetByCodePoints()**` method returns the new index after applying the offset, which is `5` in this case. It means that the character at index `5` is the desired character.
```

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
    String str = "こんにちは、世界！";
    int startIndex = 0;
    int offsetIndex = 5;

    int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
    System.out.println("New Index: " + newIndex);
  }
}
```

> The output would be:

Reviewer 2: ```suggestion
The code above results in the following output:
```

---
Title: '.offsetByCodePoints()'
Description: 'Returns the new index of a character in a string after applying the specified code point offset.'
Subjects:
  - 'Computer Science'
Tags:
  - 'Strings'
  - 'Methods'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

The **`offsetByCodePoints()`** function is used to calculate the index of a character in a string by specifying a starting index and a code point offset. It returns the new index after applying the offset limit.

It is useful when working with strings that contain characters with multiple code points, such as emojis or characters from non-Latin scripts. This function helps accurately navigate and manipulate the index based on code points, allowing you to access specific characters or portions of the string.

## Syntax

```java
Strings.offsetByCodePoints(int startIndex,int  offsetIndex)
```

- `startIndex` (int): The starting index in the string from which the offset is applied.

> 

Reviewer: ```suggestion
```
Remove extra line break.

Author: Done

- `OffsetIndex` (int): The offset, in terms of code points, by which the index is adjusted. Positive values move the index forward, and negative values move it backward.

## Example 1

In this example, the code point offset is set to 7, which means the new index will be 7. It moves the index forward by 7 code points, effectively pointing to the character `'W'` in the string.

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
  String str = "Hello, World!";
  int startIndex = 0;
  int offsetIndex = 7;

  int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
  System.out.println("New Index: " + newIndex);
  }
}
```

The output would be:

```
New Index: 7
```

> ## Example 2

Reviewer: You haven't addressed my issue with the duplicative examples:

>The next two examples are essentially the same as the first one. It might be good to show an example of fetching a character from the middle of a string of completely non-Latin script.

Author: @SSwiniarski  In Example 3, I have added a completely non-latin script.

> ## Example 2

Reviewer: Can you remove example two, as it pretty much duplicates example one.

Author: Sure

In this example, the string str contains the phrase `"hello, दुनिया!"`(meaning "hello, world!" in English). The starting index is set to 0, and the code point offset is 6.

The `**offsetByCodePoints()**` method calculates the new index by moving it forward by 6 code points. The new index will be 6, pointing to the character `'द'` (da) in the string.

```java
class OffsetByCodePoints {
  public static void main(String[] args) {
  String str = "hello, दुनिया!";
  int startIndex = 0;
  int offsetIndex = 6;

  int newIndex = str.offsetByCodePoints(startIndex, offsetIndex);
  System.out.println("New Index: " + newIndex);
  }
}
```

The output would be:

```
New Index: 6
```

## Example 3

> In this example, the string str contains the text `"Hello, 世界!".` We specify the starting index as 0 and the code point offset as 5. The `**offsetByCodePoints()**` method returns the new index after applying the offset, which is 5 in this case. It means that the character at index 5 is the desired character.

Reviewer: You haven't updated the description to match.

Author: Thank you for catching that mistake. I apologize for the confusion caused by the incorrect description. I have now updated the code and description to reflect the correct behavior.
